### PR TITLE
[Tests-Only] Api test for copying file within a public link folder

### LINKED ISSUE #338: Add more api test scenarios for copying file within public link folder
Add more api test scenarios for copying file within a public link folder. See https://github.com/owncloud/ocis-reva/issues/310

Some tests have already been added in https://github.com/owncloud/core/pull/37627. We can add tests for invalid operations to copy files/folder from the public link.

### PR DESCRIPTION
**Pull request by swoichha**
## Description
API test scenario for copying a file within within a public link folder

## Related Issue
- owncloud/ocis-reva#310
- Fixes https://github.com/owncloud/ocis-reva/issues/338

## How Has This Been Tested?
- https://github.com/owncloud/ocis/pull/367
- https://github.com/owncloud/ocis-reva/pull/351

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Review by individual-it — CHANGES_REQUESTED** [review]
more ideas what to test:
- copy a file to a name of an existing folder
- copy a file where the source name has some difficult characters
- copy a file where the destination name has some difficult characters
- copy a file where the destination folder name has some difficult characters

example for difficult characters: https://github.com/owncloud/core/blob/bcc6b5b422754e4474b901ff93047bccdde9adb9/tests/acceptance/features/apiWebdavMove1/moveFileAsync.feature#L26

- copy a folder to a destination with a name of an existing file
- copy a file/folder to itself
- copy a file/folder to an empty name

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 1]
```diff
@@ -43,3 +43,148 @@ Feature: copying from public link share
...
+    When user "Alice" copies file "/testfile.txt" to "/copy1.txt" using the new public WebDAV API
+    Then the HTTP status code should be "204"
+    And as "Alice" file "/PARENT/testfile.txt" should exist
+    And as "Alice" file "/PARENT/copy1.txt" should exist
+    And the content of file "/PARENT/copy1.txt" for user "Alice" should be "some data 0"
+
+  Scenario: Copy folder within a public link folder to same folder name as already exiting one
+    Given user "Alice" has created folder "/PARENT/testFolder"
+    And user "Alice" has uploaded file with content "some data" to "/PARENT/testFolder/testfile.txt"
+    And user "Alice" has uploaded file with content "some data" to "/PARENT/copy1.txt"
```
Comment (jasson99) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:62]: IMO the content should be different.

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 2]
```diff
@@ -43,3 +43,148 @@ Feature: copying from public link share
...
+    And user "Alice" has created a public link share with settings
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When user "Alice" copies folder "/testFolder" to "/copy1.txt" using the new public WebDAV API
```
Comment (jasson99) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:66]: is it copied to a same folder name as already existing one?

Comment (swoichha) [reply tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature]: Yes, it copies `testfile.txt` inside of folder `testFolder` to already existing file `copy1.txt`. File `copy1.txt` is overridden and it is now a folder with `testfile.txt` inside of it.

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 3]
```diff
@@ -24,22 +24,154 @@ Feature: copying from public link share
...
+    Given user "Alice" has created folder "/PARENT/testFolder"
+    And user "Alice" has uploaded file with content "some data" to "/PARENT/testFolder/testfile.txt"
+    And user "Alice" has uploaded file with content "some data 1" to "/PARENT/copy1.txt"
+    And user "Alice" has created a public link share with settings
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When the public copies folder "/testFolder" to "/copy1.txt" using the new public WebDAV API
+    Then the HTTP status code should be "204"
+    And as "Alice" folder "/PARENT/testFolder" should exist
+    And as "Alice" folder "/PARENT/copy1.txt" should exist
```
Comment (individual-it) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:69]: wow that overwrites the folder and makes it a file?

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 4]
```diff
@@ -24,22 +24,154 @@ Feature: copying from public link share
...
+    Given user "Alice" has uploaded file with content "some data" to "/PARENT/testfile.txt"
+    And user "Alice" has created folder "/PARENT/new-folder"
+    And user "Alice" has uploaded file with content "some data 1" to "/PARENT/new-folder/testfile1.txt"
+    And user "Alice" has created a public link share with settings
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When the public copies file "/testfile.txt" to "/new-folder" using the new public WebDAV API
+    Then the HTTP status code should be "204"
+    And as "Alice" file "/PARENT/testfile.txt" should exist
+    And as "Alice" folder "/PARENT/new-folder" should exist
```
Comment (individual-it) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:93]: looks like that became a file now

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 5]
```diff
@@ -24,22 +24,154 @@ Feature: copying from public link share
...
+    Then the HTTP status code should be "201"
+    And as "Alice" file "/PARENT/<destination-file-name>" should exist
+    And as "Alice" file "/PARENT/copy1.txt" should exist
+    And the content of file "/PARENT/<destination-file-name>" for user "Alice" should be "some data"
+    And the content of file "/PARENT/copy1.txt" for user "Alice" should be "some data"
+    Examples:
+      | destination-file-name |
+      | नेपाली.txt            |
+      | strängé file.txt      |
+      | C++ file.cpp          |
```
Comment (individual-it) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:113]: add a filename containing a comma, we had recently issue with that

**Review by individual-it — CHANGES_REQUESTED** [review]
added some comments, need to clarify if the overwrite is correct

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 6]
```diff
@@ -24,22 +24,158 @@ Feature: copying from public link share
...
       | permissions | read,update,create,delete |
-    When user "Alice" copies file "/testfile.txt" to "/testFolder/copy1.txt" using the new public WebDAV API
+    When the public copies file "/testfile.txt" to "/testFolder/copy1.txt" using the new public WebDAV API
     Then the HTTP status code should be "201"
     And as "Alice" file "/PARENT/testfile.txt" should exist
     And as "Alice" file "/PARENT/testFolder/copy1.txt" should exist
     And the content of file "/PARENT/testfile.txt" for user "Alice" should be "some data"
     And the content of file "/PARENT/testFolder/copy1.txt" for user "Alice" should be "some data"
+
+  Scenario: Copy file within a public link folder to same file name as already exiting one
```
Comment (phil-davis) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:47]:
```suggestion
  Scenario: Copy file within a public link folder to same file name as already existing one
```

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 7]
```diff
@@ -24,22 +24,158 @@ Feature: copying from public link share
...
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When the public copies file "/testfile.txt" to "/copy1.txt" using the new public WebDAV API
+    Then the HTTP status code should be "204"
+    And as "Alice" file "/PARENT/testfile.txt" should exist
+    And as "Alice" file "/PARENT/copy1.txt" should exist
+    And the content of file "/PARENT/copy1.txt" for user "Alice" should be "some data 0"
+
+  @skipOnOcis
+  Scenario: Copy folder within a public link folder to same folder name as already exiting one
```
Comment (phil-davis) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:60]:
```suggestion
  Scenario: Copy folder within a public link folder to the same name as an already existing file
```

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 8]
```diff
@@ -24,22 +24,158 @@ Feature: copying from public link share
...
+    And user "Alice" has created a public link share with settings
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When the public copies file "/testfile.txt" to "/copy1.txt" using the new public WebDAV API
+    And user "Alice" deletes file "/PARENT/copy1.txt" using the WebDAV API
+    Then the HTTP status code should be "204"
+    And as "Alice" file "/PARENT/copy1.txt" should not exist
+
+  @skipOnOcis
+  Scenario: Copy file within a public link folder to a file with name same as of existing folder
```
Comment (phil-davis) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:85]:
```suggestion
  Scenario: Copy file within a public link folder to a file with name same as an existing folder
```

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 9]
```diff
@@ -24,22 +24,158 @@ Feature: copying from public link share
...
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When the public copies file "/<destination-file-name>" to "/copy1.txt" using the new public WebDAV API
+    Then the HTTP status code should be "201"
+    And as "Alice" file "/PARENT/<destination-file-name>" should exist
+    And as "Alice" file "/PARENT/copy1.txt" should exist
+    And the content of file "/PARENT/<destination-file-name>" for user "Alice" should be "some data"
+    And the content of file "/PARENT/copy1.txt" for user "Alice" should be "some data"
+    Examples:
+      | destination-file-name |
```
Comment (phil-davis) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:112]:
```suggestion
      | file-name |
```
It is hard to say if this is source or destination file name - in the `Given` it is the destination to create, in the `When` it is  the source to copy from.

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 10]
```diff
@@ -24,22 +24,158 @@ Feature: copying from public link share
...
+    And as "Alice" file "/PARENT/<destination-file-name>" should exist
+    And the content of file "/PARENT/testfile.txt" for user "Alice" should be "some data"
+    And the content of file "/PARENT/<destination-file-name>" for user "Alice" should be "some data"
+    Examples:
+      | destination-file-name |
+      | नेपाली.txt            |
+      | strängé file.txt      |
+      | C++ file.cpp          |
+
+  Scenario Outline: Copy file within a public link folder to a folder with special characters
```
Comment (phil-davis) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:134]:
```suggestion
  Scenario Outline: Copy file within a public link folder into a folder with special characters
```

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 11]
```diff
@@ -24,22 +24,158 @@ Feature: copying from public link share
...
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When the public copies file "/testfile.txt" to "/<destination-file-name>/copy1.txt" using the new public WebDAV API
+    Then the HTTP status code should be "201"
+    And as "Alice" file "/PARENT/testfile.txt" should exist
+    And as "Alice" file "/PARENT/<destination-file-name>/copy1.txt" should exist
+    And the content of file "/PARENT/testfile.txt" for user "Alice" should be "some data"
+    And the content of file "/PARENT/<destination-file-name>/copy1.txt" for user "Alice" should be "some data"
+    Examples:
+      | destination-file-name |
```
Comment (phil-davis) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:147]:
```suggestion
      | destination-folder-name |
```

**Review by phil-davis — COMMENTED** [review]
We should also have ordinary COPY test scenarios that copy a file over a folder, and a folder over a file for an ordinary user's files/folders and files/folders in a received share. But that can be in a separate PR.

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature (hunk 3 above)
Comment (phil-davis) [reply tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:70]: I raised issue #37669 to add test scenarios like this for "normal" user files, and files in user/group received shares

File: config/identifier-registration.yaml [hunk 12]
```diff
@@ -0,0 +1,116 @@
+---
```
Comment (individual-it) [inline config/identifier-registration.yaml:1]: file not needed

File: ldap.crt [hunk 13]
```diff
@@ -0,0 +1,19 @@
+-----BEGIN CERTIFICATE-----
```
Comment (individual-it) [inline ldap.crt:1]: file not needed

File: ldap.key [hunk 14]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (individual-it) [inline ldap.key:1]: file not needed

File: server.crt [hunk 15]
```diff
@@ -0,0 +1,19 @@
+-----BEGIN CERTIFICATE-----
```
Comment (individual-it) [inline server.crt:1]: file not needed

File: server.key [hunk 16]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (individual-it) [inline server.key:1]: file not needed

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature [hunk 17]
```diff
@@ -24,22 +24,222 @@ Feature: copying from public link share
...
+    And user "Alice" has uploaded file with content "some data 1" to "/PARENT/new-folder/testfile1.txt"
+    And user "Alice" has created a public link share with settings
+      | path        | /PARENT                   |
+      | permissions | read,update,create,delete |
+    When the public copies file "/testfile.txt" to "/new-folder" using the new public WebDAV API
+    Then the HTTP status code should be "204"
+    And as "Alice" file "/PARENT/testfile.txt" should exist
+    And as "Alice" folder "/PARENT/new-folder" should exist
+    And the content of file "/PARENT/testfile.txt" for user "Alice" should be "some data"
+#    And the content of file "/PARENT/new-folder/testfile1.txt" for user "Alice" should be "some data 1"
```
Comment (individual-it) [inline tests/acceptance/features/apiSharePublicLink2/copyFromPublicLink.feature:130]: please remove

**Review by individual-it — CHANGES_REQUESTED** [review]
just a small comment and please squash the commits, rest looks good to me

**Review by individual-it — APPROVED** [review]

**Pull request merged**